compiler/rustc_index/src/bit_set.rs

```diff
@@ -328,6 +337,409 @@ impl<T: Idx> BitRelations<BitSet<T>> for BitSet<T> {
     }
 }
 
+/// A fixed-size bitset type with a partially dense, partially sparse
+/// representation. The bitset is broken into chunks, and chunks that are all
+/// zeros or all ones are represented and handled very efficiently.
+///
+/// This type is especially efficient for sets that typically have a large
+/// `domain_size` with significant stretches of all zeros or all ones.
```

Comment:
It feels like the added bit here that might be good is to contrast this with IntervalSet; intuitively (but tell me if you think I'm wrong I'd guess):

* ChunkedBitSet is good with *many* long stretches of zeroes or ones (e.g., domain in the tens of millions of points, hundreds of thousands of intervals each spanning several thousand points).
   * Forces *at least* (total domain) / (2048) * 24 bytes (`size_of::<Chunk>`) allocation -- for huge domain, that can already be pretty big.
* IntervalSet is optimized for arbitrarily large domain or interval size, particularly if you expect *few* intervals
   * In particular, no problem if the intervals are not distributed throughout the domain nicely.


As I write the above out, it's increasingly feeling like this adds (over IntervalSet) basically:

* Constant-time access (vs. O(log number of intervals)) to points

at the cost of:

* Up-front allocation for entire domain


Maybe worth including some parts of this in the docs -- mostly just laying out thoughts in case they seem helpful.